lib/commons/aria/roles.js

```diff
@@ -189,9 +187,151 @@ aria.implicitRole = function (node) {
+ * @method isAllowedSubsetRole
+ * @param {String} tagName tag name of a node
+ * @param {String} role aria role to check
+ * @return {Boolean} return true/false
+ * @private
+ */
+function isAllowedSubsetRole(tagName, role) {
+	if (!axe.commons.aria.lookupTable.elementAriaAllowedSubsetRolesMap[tagName]) {
+		return false;
+	}
+	return axe.commons.aria.lookupTable.elementAriaAllowedSubsetRolesMap[tagName].indexOf(role) > -1;
+};
+
+
+/**
+ * Verify if a given role is an allowed ARIA role. 
+ * @method isAllowedRole
+ * @memberof axe.commons.aria
+ * @instance
+ * @param {String} tagName tag name of a node
+ * @param {String} role aria role to check
+ * @return {Boolean} retruns true/false
+ */
+aria.isAllowedRole = function (node, tagName, role) {
+	/**
+ 	* Collection of functions, which evaluate if a given ROLE is allowed on a specified type of ELEMENT.
+ 	* @type {Object}
+ 	*/
+	const elementCheck = {
+		a: (node, role) => {
+			if (node.namespaceURI === 'http://www.w3.org/2000/svg') {
+				return true;
+			}
+			if (node.href) {
+				return isAllowedSubsetRole('aWithHref', role);
+			}
+			return true;
+		},
+		area: (node) => !(node.href),
+		button: (node, role) => {
+			if (node.getAttribute('type') === 'menu') {
+				return role === 'menuitem';
+			}
+			return isAllowedSubsetRole('button', role);
+		},
+		img: (node, role) => {
+			const hasAllowedEmptyAltRole = isAllowedSubsetRole('imgWithEmptyAlt', role);
+			if (node.alt) {
+				return !hasAllowedEmptyAltRole;
+			}
+			return hasAllowedEmptyAltRole;
+		},
+		input: (node, role) => {
+			switch (node.type) {
+			case 'button':
+				return isAllowedSubsetRole('inputTypeButton', role);
+			case 'checkbox':
+				if (role === 'button' && node.hasAttribute('aria-pressed')) {
+					return true;
+				}
+				return isAllowedSubsetRole('inputTypeCheckbox', role);
+			case 'image':
+				return isAllowedSubsetRole('inputTypeImage', role);
+			case 'radio':
+				return role === 'menuitemradio';
+			default:
+				return false;
+			}
+		},
+		li: (node, role) => {
+			const hasImplicitListitemRole = node.matches('ol li, ul li');
+			if (hasImplicitListitemRole) {
+				return isAllowedSubsetRole('li', role);
+			}
+			return true;
+		},
+		link: (node) => !(node.href),
+		menu: (node) => {
+			if (node.getAttribute('type') === 'context') {
+				return false;
+			}
+			return true;
+		},
+		option: (node) => {
+			const withinOptionList = node.matches('select > option, datalist > option, optgroup > option');
+			return !withinOptionList;
+		},
+		select: (node, role) => (!node.multiple && node.size <= 1 && role === 'menu'),
+		svg: (node, role) => {
+			// if in svg context it all roles may be used
+			if (node.parentNode && node.parentNode.namespaceURI === 'http://www.w3.org/2000/svg') {
+				return true;
+			}
+			return isAllowedSubsetRole('svg', role);
+		}
+	};
+
+	const element = axe.commons.aria.lookupTable.elementAriaAllowedRolesMap[tagName];
```

Comment:
if instead of looking that the node name matches, we include something to match on specific attributes, things will probably be significantly easier, and require far fewer special case functions like we've got in this `elementCheck` object.

Reply:
I do not want to pollute/ hoist the lookupTable in aria with this hash of functions just yet. Leaving as is.